### PR TITLE
Transmux RTC to RTMP.

### PR DESCRIPTION
**Pull request by hu-ping**
srs维护者你好：
基于srs birdger的理念实现了rtc转rtmp的功能。包括opus转码aac，h264使用jitterbuffer组帧的功能。

issue：
#2200

### CODE REVIEW

File: trunk/3rdparty/ffmpeg-4-fit/libavcodec/codec_list.c [hunk 1]
```diff
@@ -1,9 +1,11 @@
 static const AVCodec * const codec_list[] = {
     &ff_aac_encoder,
     &ff_opus_encoder,
+    &ff_libfdk_aac_encoder,
```
Comment (winlinvip) [inline trunk/3rdparty/ffmpeg-4-fit/libavcodec/codec_list.c:4]: FDK aac has a LICENSE issue, I remember it is GPL or something like that, SRS cannot use this LICENSE.
At most, SRS can only use LGPL and link it with so library. Of course, there is no problem with Apache, BSD, and MIT licenses.

`TRANS_BY_GPT3`

File: trunk/src/app/srs_app_rtc_codec.cpp [hunk 2]
```diff
@@ -130,11 +132,24 @@ srs_error_t SrsAudioDecoder::decode(SrsSample *pkt, char *buf, int &size)
...
-        // @see https://github.com/ossrs/srs/pull/2011/files
-        for (int i = 0; i < codec_ctx_->channels; i++) {
-            if (size + pcm_size * frame_->nb_samples <= max) {
-                memcpy(buf + size,frame_->data[i],pcm_size * frame_->nb_samples);
-                size += pcm_size * frame_->nb_samples;
+        if ( codec_ctx_->frame_size == 0) {
+           codec_ctx_->frame_size = frame_->nb_samples;
+        }
+
+        if (SrsAudioCodecIdAAC == codec_id_) {
```
Comment (winlinvip) [inline trunk/src/app/srs_app_rtc_codec.cpp:139]: This logic seems to be somewhat unreasonable. Does FFmpeg have a corresponding function to implement copying?

`TRANS_BY_GPT3`

File: trunk/src/app/srs_app_rtc_codec.cpp [hunk 3]
```diff
@@ -416,6 +432,10 @@ SrsAudioRecode::SrsAudioRecode(SrsAudioCodecId src_codec, SrsAudioCodecId dst_co
     dec_ = NULL;
     enc_ = NULL;
     resample_ = NULL;
+
+    fw_audio_dec = new SrsFileWriter();
```
Comment (winlinvip) [inline trunk/src/app/srs_app_rtc_codec.cpp:436]: Debug code?

`TRANS_BY_GPT3`

File: trunk/src/app/srs_app_rtc_jitbuffer.cpp [hunk 4]
```diff
@@ -1273,6 +1275,157 @@ SrsRtpFrameBufferEnum SrsRtpJitterBuffer::InsertPacket(uint16_t seq, uint32_t ts
     return buffer_state;
 }
 
+
+#if 1
+SrsRtpFrameBufferEnum SrsRtpJitterBuffer::InsertPacket2(const SrsRtpPacket2 *pkt, const char*nalu, int nalu_len,
```
Comment (winlinvip) [inline trunk/src/app/srs_app_rtc_jitbuffer.cpp:1280]: Debug code?

`TRANS_BY_GPT3`

File: trunk/src/app/srs_app_rtc_jitbuffer.cpp [hunk 5]
```diff
@@ -1475,22 +1628,27 @@ bool SrsRtpJitterBuffer::NextCompleteTimestamp(uint32_t max_wait_time_ms, uint32
...
-                } else {
-                    break;
-                }
-            } else {
-                break;
-            }
-        }
+        
+
+        //TODO:huping modify.
```
Comment (winlinvip) [inline trunk/src/app/srs_app_rtc_jitbuffer.cpp:1633]: Why is it commented out?

`TRANS_BY_GPT3`

File: trunk/src/app/srs_app_source.hpp [hunk 6]
```diff
@@ -498,6 +499,7 @@ class ISrsSourceBridger
     virtual srs_error_t on_audio(SrsSharedPtrMessage* audio) = 0;
     virtual srs_error_t on_video(SrsSharedPtrMessage* video) = 0;
     virtual void on_unpublish() = 0;
+    virtual srs_error_t on_rtp(SrsRtpPacket2* pkt) = 0;
```
Comment (winlinvip) [inline trunk/src/app/srs_app_source.hpp:502]: `on_audio/on_video` is the frame processing function for RTMP.
`on_rtp` is the packet processing function for RTP.
It is not ideal to have these two APIs together.

`TRANS_BY_GPT3`

File: trunk/3rdparty/ffmpeg-4-fit/libavcodec/codec_list.c (hunk 1 above)
Comment (hu-ping) [reply trunk/3rdparty/ffmpeg-4-fit/libavcodec/codec_list.c:4]: Is that to say that the source code of fdk aac cannot be used to compile ffmpeg, and instead we need to use the .so library to compile ffmpeg, and then call fdk aac through ffmpeg's interface?

`TRANS_BY_GPT3`

**Comment by winlinvip** [comment]
Fixed by https://github.com/ossrs/srs/tree/feature/rtc2rtmp

**Comment by winlinvip** [comment]
Fixed by #2303

**Pull request closed without merging**